### ISSUE TITLE
Sumer Release: MVP CLI commands for Sumer

### ISSUE THREAD

**Issue by bwhm** [Joystream/joystream#2149]
## Existing
Below are the current `content-directory` and `media` commands:

```
COMMANDS
  content-directory:addClassSchema              Add a new schema to a class inside content directory. Requires lead access.
  content-directory:addCuratorToGroup           Add Curator to existing Curator Group.
  content-directory:addMaintainerToClass        Add maintainer (Curator Group) to a class.
  content-directory:class                       Show Class details by id or name.
  content-directory:classes                     List existing content directory classes.
  content-directory:createClass                 Create class inside content directory. Requires lead access.
  content-directory:createCuratorGroup          Create new Curator Group.
  content-directory:createEntity                Creates a new entity in the specified class (can be executed in Member, Curator or Lead context)
  content-directory:curatorGroup                Show Curator Group details by ID.
  content-directory:curatorGroups               List existing Curator Groups.
  content-directory:entities                    Show entities list by class id or name.
  content-directory:entity                      Show Entity details by id.
  content-directory:initialize                  Initialize content directory with input data from @joystream/content library or custom, provided one. Requires lead access.
  content-directory:removeCuratorFromGroup      Remove Curator from Curator Group.
  content-directory:removeCuratorGroup          Remove existing Curator Group.
  content-directory:removeEntity                Removes a single entity by id (can be executed in Member, Curator or Lead context)
  content-directory:removeMaintainerFromClass   Remove maintainer (Curator Group) from class.
  content-directory:setCuratorGroupStatus       Set Curator Group status (Active/Inactive).
  content-directory:updateClassPermissions      Update permissions in given class.
  content-directory:updateEntityPropertyValues  Updates the property values of the specified entity (can be executed in Member, Curator or Lead context)

COMMANDS
  media:createChannel       Create a new channel on Joystream (requires a membership).
  media:curateContent       Set the curation status of given entity (Channel/Video). Requires Curator access.
  media:featuredVideos      Show a list of currently featured videos.
  media:myChannels          Show the list of channels associated with current account's membership.
  media:myVideos            Show the list of videos associated with current account's membership.
  media:removeChannel       Removes a channel (required controller access).
  media:removeVideo         Remove given Video entity and associated entities (VideoMedia, License) from content directory.
  media:setFeaturedVideos   Set currently featured videos (requires lead/maintainer access).
  media:updateChannel       Update one of the owned channels on Joystream (requires a membership).
  media:updateVideo         Update existing video information (requires controller/maintainer access).
  media:updateVideoLicense  Update existing video license (requires controller/maintainer access).
  media:uploadVideo         Upload a new Video to a channel (requires a membership).
```

## MVP commands for Sumer
I assume there will not be support for episodic content nor playlists.

Instead of a `content-directory` and `media`, I propose either `content-curators` and `media` to separate between them, or use `media` for all now that there is an alternative for "regular users" to create, update and manage content.

### content-curators
All users can of course query, but making changes here requires permissions outside of a membership.

#### initialize
To initialize the available Video/Channel Categories, Languages, Person, etc. with ids.

#### class, classes, entity, entities
As there are no global concept of `classes`, `entities` and `schemas` in this content directory, these commands will need to be renamed. However, there must be a way to read what is currently in there. Note that this might not be the most intuitive commands. Feel free to come up with better command names, and/or query structure. Not sure if an `itemType` command is needed, so excluding it here.

Examples:

#### items
`joystream-cli content-directory:items videoCategory`
returns all the available Video Categories, and their id.

#### item
`joystream-cli content-directory:item --videoCategory 3`
returns the Video Category with id 3

`joystream-cli content-directory:item --videoCategory Comedy`
returns the id of the Video Category named Comedy

#### createItem
To add a new Video Category - Comedy, or similar.
`joystream-cli content-directory:createItem --videoCategory Comedy`

#### createCuratorGroup, addCuratorToGroup, removeCuratorFromGroup, removeCuratorGroup
All of these are needed to manage curators, and their permissions.

#### curatorGroup, curatorGroups
These are needed to get information about curators groups, and their permissions.

#### curateChannel, curateVideo
Replaces `curateContent`.

#### deleteItem
Deletes a Video, Channel, Video Category, etc.
Not sure if/how this is implemented on the runtime.
As it's a far more serious action than censoring, it should perhaps require a different set of permissions?

#### setFeaturedVideos
To set the features videos.

#### featuredVideos
To see the features videos.

### media
These are all commands a channel owner/controller can perform.

#### createChannel
Create a new channel.
Must allow json input for bulk creation.

#### myChannels
To see all the channels you control.

#### myVideos
To see all the videos you have made (under a channel?).

#### removeChannel
Deletes a channel, and all videos it has.

#### removeVideo
Deletes a video.

#### updateChannel
To update one of your channels.

#### updateVideo
To update one of your videos.

#### uploadVideo
Upload a new video to a channel you own.
Must allow json input for bulk uploads.

**Comment by bwhm** [comment]
@iorveth 
@ondratra 

This is what I would consider the MVP commands. Some are still more urgent than others, but removing some of them will mean a user or curator would need to use Atlas and/or extrinsics to complete their tasks.

As this list is made to create issues and give estimates, it's not the end of the world if something is wrong/missing, but perhaps @mnaamani could still verify?

**Comment by bedeho** [comment]
1. I do not think this is sufficiently MVP, for example, all of the commands here having to do with channel creation, video creation and removal and management, all the item stuff for reading, why is that needed for Sumer to be released? I think, at worst, whatever the initialization script may need could be added, but I think there should be room to cut this down even more.
2. I think this description goes too far in describing the CLI UX, but does that badly (1), but then in other ways seems to underspecify this (2). For example for (1) I would say almost the entire way of interacting should be changed here. There is no such thing as a item, or creatItem. The weird camelCase way of writing comands is not something I have hardly seen, it does not look effective to me. People are used to doing `git remote add ..` or `git remote remove ..` where `remote` becomes a sort of logical namespace for actions, rather than `git addRemote `... About (2) it does not go into for example what sort of help or validation should or should not be done on assets (image size, aspect ratio, encodings, etc.) or input values. 
3. I do not know fully what is meant with `initialize`, but it sounds like something which should not be part of this CLI itself.

In summary, I think the CLI dev here should have the responsibility to come up with a concrete command and interaction layout that addresses the MVP functionality required for Sumer, and this functionality should come from @bwhm.

**Comment by bwhm** [comment]
> I do not think this is sufficiently MVP, for example, all of the commands here having to do with channel creation, video creation and removal and management, all the item stuff for reading, why is that needed for Sumer to be released?

Wrt channel creation and video uploads, we were in agreement that batch creation/uploads are needed. The way it's currently done is using the single creation/upload command with a json input. That way, a bash script can be used for batching. We need batching -> we need this command.

Removal and updating can arguably be removed.

Reading is to keep the ability to use the cli without having to simultaneously check in Pioneer/Atlas.

>  I think, at worst, whatever the initialization script may need could be added, but I think there should be room to cut this down even more.

I don't understand this part.

> I think this description goes too far in describing the CLI UX, but does that badly (1), but then in other ways seems to underspecify this (2). For example for (1) I would say almost the entire way of interacting should be changed here.

I have of course used the current CLI as the starting point here, and assumed the person implementing this will look at this for reference.

> There is no such thing as a item, or creatItem. 

As stated above, I agree `item` is not good. As we are no longer using the concept of classes and entities, I chose this as a generic term for explaining what it does.

> The weird camelCase way of writing comands is not something I have hardly seen, it does not look effective to me. People are used to doing git remote add .. or git remote remove .. where remote becomes a sort of logical namespace for actions, rather than git addRemote ... About (2) it does not go into for example what sort of help or validation should or should not be done on assets (image size, aspect ratio, encodings, etc.) or input values.

This is how the CLI is now. If we change this, we need to change the entire CLI, which is arguably a bigger task than just removing deprecated commands and adding new ones for the MVP.

> I do not know fully what is meant with initialize, but it sounds like something which should not be part of this CLI itself.

We could of course make this as a script instead. I assume it was put in to the CLI to make dev chains easier.

> In summary, I think the CLI dev here should have the responsibility to come up with a concrete command and interaction layout that addresses the MVP functionality required for Sumer, and this functionality should come from @bwhm 

I don't mind doing that, but we at least have to agree on the scope. 
1. Are we making the CLI work with Sumer, or are we doing a complete overhaul?
2. Are we doing an MVP in terms of adding features, but target a "better" UX than before?
3. Is it insufficient to refer to current implementation when replacing an existing/similar command, or do I need to add the full details?

**Comment by bedeho** [comment]
1. I was thinking we could change the cli ux just for the content directory, but now that I think of it it would be stupid to have it mixed, and its too much work to fix it all now, so lets just continue with the overall command structure as is then.
2. Initialization, that is in this case really just category creation of some set categories, is needed as part of Sumer deployment, and for the dev deployment, but I do not think it should be part of this CLI.
3. We agree to drop the item approach.
4. Input validation still needs to be addressed, and I think we do need it in quite a few places, informed by the protobuffer standard.

**Comment by bwhm** [comment]
Keeping the original post for clarity, but consider this the commands:

### Commands for channel owners

#### media:createChannel
Create a new channel. Must include input validation.

In the current implementation, this command can take a json as input, or provide an interactive experience. As Atlas will contain a far better version of the latter, I'd say it's sufficient to only allow a json as input in the CLI.

Use current implementation as reference, but change the json structure where applicable to match the inputs. Should also include examples in `--help` to assist in json creation.

#### media:uploadVideo
Uploads a video to the content directory.

In the current implementation, this command can take a json as input, or provide an interactive experience. As Atlas will contain a far better version of the latter, I'd say it's sufficient to only allow a json as input in the CLI.

Use current implementation as reference, but change the json structure where applicable to match the inputs. Should also include examples in `--help` to assist in json creation.


#### media:updateChannel
#### media:updateVideo
Neither of these are strictly necessary for an MVP, as updating can be done in Atlas. OTOH, they should be somewhat simple once the commands above are implemented.

I'd say we skip them.


### Commands for curators, curator groups and permissions
All the commands below are needed for `curators` to perform their tasks.

#### content-directory:createCuratorGroup
Creates a new `curatorGroup`.
Use current implementation as reference.

#### content-directory:removeCuratorGroup
Removes an existing `curatorGroup`, revoking any permissions the `curators` in the group has.
Use current implementation as reference.

#### content-directory:addCuratorToGroup
Adds a `curator` to a `curatorGroup`.
Use current implementation as reference.

#### content-directory:removeCuratorFromGroup
Removes a `curator` from a `curatorGroup`.
Use current implementation as reference.

#### content-directory:curatorGroup
Lists information about a `curatorGroup` by its id.
Use current implementation as reference.

#### content-directory:setFeaturedVideos
To set the features videos.
Should work like the current implementation, as described below.
```
USAGE
  $ joystream-cli media:setFeaturedVideos VIDEOIDS

ARGUMENTS
  VIDEOIDS  Comma-separated video ids

OPTIONS
  --add  If provided - currently featured videos will not be removed.
```

#### content-directory:featuredVideos
Returns a list of the features videos in a table format:

|Video ID     |Video Title  |Channel ID   |Channel Title|
|-------------|-------------|-------------|-------------|
|1            |Ghost Signals|1            |Schism       |


### Create metadata entries
Not entirely sure what the best option is here long term, but as long as we only have a limited set of types at launch, it's ok to make individual commands for each, meaning all of the below commands:

#### content-directory:createMediaType
Creates a new media type.

Example:
```
USAGE
  $ joystream-cli content-directory:createMediaType

OPTIONS
  -i, --input=input      Path to JSON file to use as input (if not specified - the input can be provided interactively)
  -y, --confirm          Confirm the provided input
```

#### content-directory:createVideoCategory
Creates a new video category.

#### content-directory:createChannelCategory
Creates a new channel category.

#### content-directory:createLanguage
Creates a new language.

#### content-directory:createLicense
Creates a new license.

All of these would work the same way as shown in the `createMediaType` example.

Depending on the implementation, it may be necessary to provide a `curatorId` and/or `curatorGroupId` as an argument.

---

The alternative here would be something like `content-directory:createMetadataEntry` as outlined below:

#### content-directory:createMetadataEntry
Which takes a `type` argument:

```
USAGE
  $ joystream-cli content-directory:createMetadataEntry TYPE

ARGUMENTS
  TYPE      Eg. language, license, mediaType, videoCategory, channelCategory

OPTIONS
  -i, --input=input      Path to JSON file to use as input (if not specified - the input can be provided interactively)
  -y, --confirm          Confirm the provided input
```

Depending on the implementation, it may be necessary to provide a `curatorId` and/or `curatorGroupId` as an argument.

Long term, I think this is a better solution, but still struggling with creating a command name that is "correct" and intuitive. Feel free to propose something else than `createMetadataEntry`.

**Comment by mnaamani** [comment]
## Channel ownership

It should be stated that we currently support two "Actors" that can be channel owners: Members and CuratorGroups. Any curator in a curator group can act as the channel owner of a Channel owned by that group.
So the command to create/update channels and videos should support these two actor types.

The basic creation of channels and videos is sufficient for an mvp but I agree adding support to update is not too much additional work.

## Language, License, MediaType

In the new content directory `Language`, `Licence`, `MediaType` and other metadata values are no longer defined in the content directory itself by lead and curators, but are "opaque" values. We just follow a standard which all clients must follow to make the values meaningful. Details can be found in the [protobuf docs](https://github.com/Joystream/joystream/blob/sumer/content-metadata-protobuf/doc/index.md)

For example:

Language will be an `ISO_639-1` string code:  for English: `'en'`
License will be a numeric code defined by us. (details below)

So we will not need:
`content-directory:createLanguage` - see `language` field in [ChannelMetadata](https://github.com/Joystream/joystream/blob/sumer/content-metadata-protobuf/doc/index.md#channelmetadata) message
`content-directory:createMediaType`:  - see [MediaType](https://github.com/Joystream/joystream/blob/sumer/content-metadata-protobuf/doc/index.md#mediatype) message
`content-directory:createLicense`: - see [License](https://github.com/Joystream/joystream/blob/sumer/content-metadata-protobuf/doc/index.md#license) message
`content-directory:createMetadataEntry`

Leaving us with:
`content-directory:createVideoCategory`
`content-directory:createChannelCategory`
We should also add their `delete` counter-part.

## Curators
The list of commands you provided is fine, (not some extrinsics related to curator groups management have been slightly renamed (delete instead of remove)

We should also add
`content-directory:censorVideo`
`content-directory:censorChannel`
`content-directory:uncensorVideo`
`content-directory:uncensorChannel`
